### DISCUSSION TITLE
Better mod folder/dat ordering

### DISCUSSION THREAD

**Discussion by phobos2077** [sfall-team/sfall#450] [Ideas]
After seeing how different mod authors approach this issue differently, I thought there should be a solution that will be convenient for everyone. 

* Auto-loading DAT files has the issue that you need to fiddle with file names to properly order mods.
* PatchFileXX options in ddraw.ini are also bad because you need to constantly tweak these numbers to reorder mods, it's just extra work for the user.

Proposed solution:
* We add separate mods/mod_order.txt file.
* Sfall will read entries from this file and add them to the end of the patch load list.
* Auto-loading of DAT files is kept, for backwards compatibility, IF mod_order.txt is empty or missing.

This should make manual installation of mods easier and also guard against mod conflicts by promoting proper mod ordering.

In a perfect world, I would like to see some kind of mod manager (maybe there is some simple open source one we can fork for Fallout) and all of the mods placed on github, with a kind of package.json file, describing meta info about the mod (version, dependencies, etc.). This will completely eliminate any need for mod repacks, as modding Fallout will be as easy as any modern game.

**Comment by burner1024** [comment]
I think it's a little too optimistic.
Repacks aren't going anywhere, as the reasons for their existence are 50% people laziness and 50% stroking repackers egoes.
And ordering is just moving the issue from one place to another.

Still, it would be nice to have a simpler ordering system, and a mod manager, but I don't think it's really going to change things.
What I'd like to see is more possibilities for making mods compatible in a way that load order wouldn't matter, and better guidelines on how to do that.

**Reply by phobos2077** [reply]
Well pretty much all games I've used mods with have load ordering issues, there's no magic pill against it. It's a fundamental issue where you can't guarantee that mod A won't change the same thing that mod B does. So we still need ordering. What I'm proposing is one step into what I believe to be the right direction.

**Comment by phobos2077** [comment]
Ok, silly of me, I've discovered this ~~shit show of an~~ issue https://github.com/sfall-team/sfall/issues/206.

After reading through all that and trying both @Lexx2k and @burner1024 approaches, I still think my suggestion would solve exactly what Lexx was criticizing about RPU approach. I want for mod order to not be dependent on file names - this is just pure insanity. I don't really know of any example of a game with some kind of  modding support that does ordering based on names. 

Also, this solves another issue I'm having of easily disabling mods. I tried to use Mod Organizer 2 with a custom Fallout 2 plugin for this, but I think this load_order.txt will achieve mostly the same result. And unlike with MO2 (or until we implement a proper plugin for it), it's very straightforward for mod users and certainly better than "PatchFileXXX" nonsense with EtTu mod.

**Comment by NovaRain** [comment]
>"PatchFileXXX" nonsense

In my opinion, the [ExtraPatches] section is merely an analog to the `load_order.txt`, with the XX numbers work as line numbers. It's just like an embedded `load_order.txt` in ddraw.ini.

My idea for the `load_order.txt`:
* Leave the existing [ExtraPatches] and .dat autoloading mechanism be. Just add yet another section to the data load order chain, with the entries loaded from `load_order.txt` have even higher priority than the other two.
* Check whether the `load_order.txt` exists. If yes, then load entries from it and ignore the other two. If no, then fall back to how things currently work (.dat autoloading + [ExtraPatches]).
* Or, if the `load_order.txt` doesn't exist, create one from the result of .dat autoloading and [ExtraPatches]. In the first time, the game uses the data load order chain from .dat autoloading + [ExtraPatches] and create the list, on the next game launch it will use the list instead. Kinda overdesign I guess.

**Comment by phobos2077** [comment]
FYI mods_order.txt system got implemented in develop branch. It auto-creates the file based on mods folder contents the first time (when file doesn't exist yet), then it just uses the load order from the file without any "smart" logic.
